File: aleam/polars_integration.py
"""
Polars integration for Aleam.
Provides true random data for Polars DataFrames.
"""

import polars as pl
import numpy as np
from typing import Optional, Union, List, Any
from .core import Aleam, AleamBase


class PolarsGenerator:
    """
    Polars-compatible random generator using true randomness.
    
    Example:
        >>> from aleam.polars_integration import PolarsGenerator
        >>> gen = PolarsGenerator()
        >>> df = gen.dataframe(100, columns=['a', 'b', 'c'])
    """
    
    def __init__(self, rng: Optional[AleamBase] = None):
        self.rng = rng or Aleam()
    
    def series(self, n: int, distribution: str = "uniform",
               params: dict = None, name: str = None) -> pl.Series:
        """Generate true random Polars Series"""
        params = params or {}
        
        if distribution == "uniform":
            low = params.get('low', 0.0)
            high = params.get('high', 1.0)
            data = [self.rng.uniform(low, high) for _ in range(n)]
        elif distribution == "normal":
            mu = params.get('mu', 0.0)
            sigma = params.get('sigma', 1.0)
            data = [self.rng.gauss(mu, sigma) for _ in range(n)]
        elif distribution == "exponential":
            rate = params.get('rate', 1.0)
            data = [self.rng.exponential(rate) for _ in range(n)]
        elif distribution == "poisson":
            lam = params.get('lam', 1.0)
            data = [self.rng.poisson(lam) for _ in range(n)]
        elif distribution == "choice":
            choices = params.get('choices', [0, 1])
            data = [self.rng.choice(choices) for _ in range(n)]
        else:
            raise ValueError(f"Unknown distribution: {distribution}")
        
        return pl.Series(name, data)
    
    def dataframe(self, n: int, columns: List[str],
                  distributions: Optional[dict] = None) -> pl.DataFrame:
        """Generate true random Polars DataFrame"""
        if distributions is None:
            distributions = {col: ('uniform', {}) for col in columns}
        
        data = {}
        for col in columns:
            if col in distributions:
                dist, params = distributions[col]
            else:
                dist, params = 'uniform', {}
            
            data[col] = self.series(n, dist, params, name=col)
        
        return pl.DataFrame(data)
    
    def shuffle(self, df: pl.DataFrame) -> pl.DataFrame:
        """Shuffle DataFrame rows with true randomness"""
        indices = list(range(df.height))
        self.rng.shuffle(indices)
        return df[indices]

File: src/aleam/integrations/polars_integration.cpp
/**
 * @file polars_integration.cpp
 * @brief Polars integration implementation for Aleam
 * @license MIT
 */

#include "polars_integration.h"
#include <cmath>
#include <algorithm>
#include <numeric>
#include <unordered_map>
#include <sstream>

namespace aleam {
namespace integrations {

PolarsGenerator::PolarsGenerator(AleamCore& rng)
    : m_rng(rng) {
}

std::unordered_map<std::string, double> PolarsGenerator::parse_params(const std::string& params) {
    std::unordered_map<std::string, double> result;
    if (params.empty()) return result;
    
    std::istringstream iss(params);
    std::string token;
    
    while (std::getline(iss, token, ',')) {
        size_t eq_pos = token.find('=');
        if (eq_pos != std::string::npos) {
            std::string key = token.substr(0, eq_pos);
            double value = std::stod(token.substr(eq_pos + 1));
            result[key] = value;
        }
    }
    
    return result;
}

std::vector<double> PolarsGenerator::generate_uniform(size_t n, double low, double high) {
    std::vector<double> result(n);
    double range = high - low;
    
    for (size_t i = 0; i < n; ++i) {
        result[i] = low + m_rng.random() * range;
    }
    
    return result;
}

std::vector<double> PolarsGenerator::generate_normal(size_t n, double mu, double sigma) {
    std::vector<double> result(n);
    
    for (size_t i = 0; i < n; ++i) {
        double u1 = m_rng.random();
        double u2 = m_rng.random();
        double z = std::sqrt(-2.0 * std::log(u1)) * std::cos(2.0 * M_PI * u2);
        result[i] = mu + sigma * z;
    }
    
    return result;
}

std::vector<double> PolarsGenerator::generate_exponential(size_t n, double rate) {
    std::vector<double> result(n);
    
    for (size_t i = 0; i < n; ++i) {
        result[i] = -std::log(1.0 - m_rng.random()) / rate;
    }
    
    return result;
}

std::vector<double> PolarsGenerator::series_f64(size_t n, const std::string& distribution, 
                                                  const std::string& params) {
    auto parsed = parse_params(params);
    
    if (distribution == "uniform") {
        double low = parsed.count("low") ? parsed["low"] : 0.0;
        double high = parsed.count("high") ? parsed["high"] : 1.0;
        return generate_uniform(n, low, high);
    } else if (distribution == "normal") {
        double mu = parsed.count("mu") ? parsed["mu"] : 0.0;
        double sigma = parsed.count("sigma") ? parsed["sigma"] : 1.0;
        return generate_normal(n, mu, sigma);
    } else if (distribution == "exponential") {
        double rate = parsed.count("rate") ? parsed["rate"] : 1.0;
        return generate_exponential(n, rate);
    } else {
        // Default to uniform
        return generate_uniform(n, 0.0, 1.0);
    }
}

std::vector<int64_t> PolarsGenerator::series_i64(size_t n, int64_t low, int64_t high) {
    std::vector<int64_t> result(n);
    int64_t range = high - low;
    
    for (size_t i = 0; i < n; ++i) {
        result[i] = low + static_cast<int64_t>(m_rng.random() * range);
        if (result[i] >= high) result[i] = high - 1;
    }
    
    return result;
}

std::vector<bool> PolarsGenerator::series_bool(size_t n, double p) {
    std::vector<bool> result(n);
    
    for (size_t i = 0; i < n; ++i) {
        result[i] = (m_rng.random() < p);
    }
    
    return result;
}

std::vector<size_t> PolarsGenerator::shuffle_indices(size_t n) {
    std::vector<size_t> indices(n);
    std::iota(indices.begin(), indices.end(), 0);
    
    // Fisher-Yates shuffle
    for (size_t i = n - 1; i > 0; --i) {
        size_t j = static_cast<size_t>(m_rng.random() * (i + 1));
        if (j > i) j = i;
        std::swap(indices[i], indices[j]);
    }
    
    return indices;
}

std::vector<std::vector<double>> PolarsGenerator::dataframe_f64(size_t rows, size_t columns,
                                                                  const std::string& distribution) {
    std::vector<std::vector<double>> result(rows, std::vector<double>(columns));
    
    for (size_t j = 0; j < columns; ++j) {
        auto col_data = series_f64(rows, distribution);
        for (size_t i = 0; i < rows; ++i) {
            result[i][j] = col_data[i];
        }
    }
    
    return result;
}

std::vector<std::vector<int64_t>> PolarsGenerator::dataframe_i64(size_t rows, size_t columns,
                                                                   int64_t low, int64_t high) {
    std::vector<std::vector<int64_t>> result(rows, std::vector<int64_t>(columns));
    
    for (size_t j = 0; j < columns; ++j) {
        auto col_data = series_i64(rows, low, high);
        for (size_t i = 0; i < rows; ++i) {
            result[i][j] = col_data[i];
        }
    }
    
    return result;
}

}  // namespace integrations
}  // namespace aleam